File: turing_machine.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define N 1000 // Tamaño de la cinta

typedef struct {
    char *tape;   // Cinta de la máquina
    int head;     // Posición de la cabeza
    char state[10]; // Estado actual
    char trf[100][5][5]; // Tabla de estados internos
} TuringMachine;

// Función para inicializar la máquina de Turing
void initTuringMachine(TuringMachine *tm, const char *algo, const char *input, int state) {
    tm->tape = (char *)malloc(N * sizeof(char));
    memset(tm->tape, '_', N); // Inicializar la cinta con guiones bajos
    tm->head = N / 2; // Colocar la cabeza en el centro de la cinta
    strncpy(tm->tape + tm->head, input, strlen(input)); // Insertar el input en el centro de la cinta
    snprintf(tm->state, sizeof(tm->state), "%d", state); // Establecer el estado inicial

    printf("\nInternal states table\n---------------------\n");

    // Leer y asignar los valores de la tabla de estados internos
    int line_num = 0;
    char line[256];
    const char *delim = ",";
    char *token;
    char algo_copy[strlen(algo) + 1];
    strcpy(algo_copy, algo);

    char *line_ptr = strtok(algo_copy, "\n");
    while (line_ptr != NULL) {
        strcpy(line, line_ptr);
        token = strtok(line, delim);
        strcpy(tm->trf[line_num][0], token); // Estado actual

        for (int i = 1; i < 5; ++i) {
            token = strtok(NULL, delim);
            strcpy(tm->trf[line_num][i], token); // Leer, escribir, dirección, próximo estado
        }

        printf("%s\n", line_ptr);
        line_ptr = strtok(NULL, "\n");
        line_num++;
    }
    printf("---------------------\n");
}

// Función para realizar un solo paso en la cinta
void shiftOneStep(TuringMachine *tm, int i) {
    if (strcmp(tm->state, "H") != 0) { // Si no está en estado de Halt (H)
        char a = tm->tape[tm->head]; // Leer valor en la posición actual de la cabeza
        for (int j = 0; j < 100; j++) { // Buscar la transición correspondiente
            if (strcmp(tm->trf[j][0], tm->state) == 0 && tm->trf[j][1][0] == a) {
                char r = tm->trf[j][2][0]; // Valor a escribir
                char d = tm->trf[j][3][0]; // Dirección (r, l, o *)
                strcpy(tm->state, tm->trf[j][4]); // Próximo estado

                tm->tape[tm->head] = r; // Escribir en la cinta

                if (d == 'r') {
                    tm->head++;
                } else if (d == 'l') {
                    tm->head--;
                }
                
                printf("%d %s %s\n", i + 1, tm->tape, tm->state);
                printf("---------\n");
                break;
            }
        }
    }
}

// Función para ejecutar la máquina de Turing
void execute(TuringMachine *tm, const char *first_bin_num, const char *second_bin_num, int max_iter) {
    printf("\ni in   state\n------------\n");
    int i = 0;
    while (strcmp(tm->state, "H") != 0 && i < max_iter) {
        shiftOneStep(tm, i);
        i++;
    }

    // Mostrar el resultado
    char res[1000];
    strncpy(res, tm->tape, strlen(tm->tape));
    for (int j = 0; j < strlen(res); j++) {
        if (res[j] == '_') {
            res[j] = '\0'; // Limpiar guiones bajos de la cinta
        }
    }
    printf("\nRESULT\n");
    printf("%s + %s = %s (binary)\n", first_bin_num, second_bin_num, res);
    printf("%d + %d = %d (decimal)\n", (int)strtol(first_bin_num, NULL, 2), (int)strtol(second_bin_num, NULL, 2), (int)strtol(res, NULL, 2));
}

int main() {
    // Solicitar input al usuario
    printf("Adding two binary numbers x + y\n");
    char first_bin_num[100], second_bin_num[100];
    printf("x = ");
    scanf("%s", first_bin_num);
    printf("y = ");
    scanf("%s", second_bin_num);
    
    char input[200];
    snprintf(input, sizeof(input), "%s_%s", first_bin_num, second_bin_num);

    // Leer la tabla de estados internos desde el archivo
    FILE *file = fopen("add_2_binary_numbers_2.txt", "r");
    if (!file) {
        perror("Error al abrir el archivo.");
        return 1;
    }
    
    char algo[1000];
    fread(algo, sizeof(char), 1000, file);
    fclose(file);

    // Inicializar y ejecutar la máquina de Turing
    TuringMachine tm;
    initTuringMachine(&tm, algo, input, 0);
    execute(&tm, first_bin_num, second_bin_num, 10000);

    // Liberar memoria
    free(tm.tape);

    return 0;
}

File: turing_machine.py
import set_internal_states_table as sist  # Importación correcta

N = 1000  # longitud de la cinta, inicializada con un valor grande

class MaquinaDeTuring:
    def __init__(self, algoritmo, entrada, estado=0):
        # inicializar diccionario
        self.transiciones = {}
        # convertir el estado a cadena
        self.estado = str(estado)
        # inicializar la cinta con guiones bajos
        self.cinta = ['_'] * N
        # establecer la posición de la cabeza
        self.cabeza = N // 2  # la cabeza está en el medio
        # insertar la entrada en el medio de la cinta
        for i, char in enumerate(entrada):
            self.cinta[self.cabeza + i] = char

        print('\nTabla de estados internos')
        print('-------------------------')
        # asignar valor a la clave
        for linea in algoritmo.splitlines():  # por cada línea en la tabla de estados internos
            # s, s1: estado actual y siguiente estado; d: dirección de desplazamiento;
            s, a, r, d, s1 = linea.split(',')  # a: valor leído; r: valor a escribir
            self.transiciones[s, a] = (r, d, s1)  # estado actual y valor a aceptar
            print(linea)
            print('---------')

    def mover_un_paso(self, i):
        # H significa detenerse
        if self.estado != 'H':
            # obtener el valor leído en la posición actual de la cinta
            valor_leido = self.cinta[self.cabeza]
            # obtener la acción basada en el estado actual y el valor leído
            accion = self.transiciones.get((self.estado, valor_leido))

            if accion:  # si hay una acción (r, d, s1)
                # asignar r = valor a escribir, d = dirección, s1 = siguiente estado
                valor_a_escribir, direccion, siguiente_estado = accion
                # escribir el valor en la cinta
                self.cinta[self.cabeza] = valor_a_escribir
                # mover la cabeza y cambiar el estado
                if direccion != '*':
                    self.cabeza += 1 if direccion == 'r' else -1
                self.estado = siguiente_estado
                print(str(i + 1) + ' ' + ''.join(self.cinta).replace('_', ''), self.estado)
                print('---------')

    def ejecutar(self, max_iter=10000, primer_num_bin=None, segundo_num_bin=None):
        print('\ni en   estado')
        print('------------')
        i = 0
        # ejecutar la máquina hasta que se detenga o llegue al máximo de iteraciones
        while self.estado != 'H' and i < max_iter:  # prevenir bucles infinitos
            self.mover_un_paso(i)
            i += 1
        # limpiar la cinta
        resultado = ''.join(self.cinta).replace('_', '')
        print('\nRESULTADO')
        # salida de los resultados
        print(f"{primer_num_bin} + {segundo_num_bin} = {resultado} (binario)")
        print(f"{int(primer_num_bin, 2)} + {int(segundo_num_bin, 2)} = {int(resultado, 2)} (decimal)")


if __name__ == '__main__':
    # encabezado
    print("Suma de dos números binarios x + y")
    # recibir entradas
    primer_num_bin = str(input("x = "))
    segundo_num_bin = str(input("y = "))
    entrada = primer_num_bin + '_' + segundo_num_bin
    # leer la tabla de estados internos
    with open(sist.sum_of_two_binary_numbers) as f:
        algoritmo = f.read()
    # ejecutar la máquina de Turing
    turing = MaquinaDeTuring(algoritmo, entrada)
    turing.ejecutar(primer_num_bin=primer_num_bin, segundo_num_bin=segundo_num_bin)
